File: Linear Regression/MultiLinear.py
from matplotlib import pyplot as pp
from mpl_toolkits.mplot3d import Axes3D
import numpy as np
import pandas as pd
from matplotlib.animation import FuncAnimation

# Plot
def plotresults(theta):
    x1 = np.linspace(X["Te"].min(), X["Te"].max(), 100)
    x2_fixed = np.mean(X["WS"])
    
    y_pred = theta[0] + theta[1] * x1 + theta[2] * x2_fixed
    
    fig = pp.figure()
    ax = fig.add_subplot(111, projection="3d")
    
    ax.plot(x1, np.full_like(x1, x2_fixed), y_pred, color='blue')
    
    ax.scatter(X["Te"], X["WS"], y, color='red')
    def update(angle):
        ax.view_init(elev=20, azim=angle)

    # Animate rotation
    ani = FuncAnimation(fig, update, frames=np.arange(0, 360, 2), interval=50)

    pp.show()

def linear_regression(X, y, lr=0.01, epochs=10000, dec=0.9999, catch=1e-10, min_lr=1e-10):
    np.random.seed(60) 
    m, n = X.shape
    X = np.c_[np.ones(m), X] 
    theta = np.random.randn(n + 1)
    mse = float("inf")

    for epoch in range(epochs):
        predictions = X @ theta
        errors = predictions - y
        gradient = (1/m) * X.T @ errors
        theta -= lr * gradient
        lr = max(lr * dec, min_lr)

        new_mse = np.mean(errors ** 2)
        if epoch % 10000 == 0:
            print(f"Epoch {epoch}: MSE = {new_mse}")

        if abs(mse - new_mse) < catch:
            break 
        mse = new_mse
    
    return theta

 
# Load data
raw_data = pd.read_csv("./Linear Regression/WeatherDataM.csv")
X = raw_data[["Te", "WS"]]
y = raw_data.Hu
 
# Train model
theta = linear_regression(X, y, lr=0.0001, epochs=500000, dec=1)
print("Theta:", theta)
 
plotresults(theta)

File: Linear Regression/multilinear.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cmath>

using namespace std;

typedef vector<double> Vec;
typedef vector<Vec> Mat;

class LinearRegression {
private:
    Vec theta;
    double lr;
    double dec;
    double catch_threshold;
    double min_lr;
    int epochs;

public:
    LinearRegression(double learning_rate = 0.01, int max_epochs = 10000, double decay = 0.9999, double catch_thr = 1e-6, double min_lr = 1e-8)
        : lr(learning_rate), epochs(max_epochs), dec(decay), catch_threshold(catch_thr), min_lr(min_lr) {}

    Vec fit(const Mat &X, const Vec &y) {
        int m = X.size();
        int n = X[0].size();
        theta.resize(n + 1);
        random_device rd;
        mt19937 gen(rd());
        normal_distribution<> d(0, 1);

        for (double &t : theta) t = d(gen); // Random initialization

        double mse = numeric_limits<double>::infinity();

        for (int epoch = 0; epoch < epochs; epoch++) {
            Vec predictions(m, 0.0);
            for (int i = 0; i < m; i++) {
                predictions[i] = theta[0]; // Bias term
                for (int j = 0; j < n; j++) {
                    predictions[i] += theta[j + 1] * X[i][j];
                }
            }

            Vec errors(m);
            for (int i = 0; i < m; i++) errors[i] = predictions[i] - y[i];

            Vec gradient(n + 1, 0.0);
            for (int j = 0; j < n + 1; j++) {
                for (int i = 0; i < m; i++) {
                    gradient[j] += (j == 0 ? 1.0 : X[i][j - 1]) * errors[i];
                }
                gradient[j] /= m;
            }

            for (int j = 0; j < n + 1; j++) {
                theta[j] -= lr * gradient[j];
            }

            lr = max(lr * dec, min_lr);

            double new_mse = 0.0;
            for (double err : errors) new_mse += err * err;
            new_mse /= m;

            if (epoch % 1000 == 0) {
                cout << "Epoch " << epoch << ": MSE = " << new_mse << endl;
            }

            if (abs(mse - new_mse) < catch_threshold) {
                break; // Early stopping
            }
            mse = new_mse;
        }
        return theta;
    }
};

int main() {
    Mat X = {{1, 2}, {2, 3}, {3, 4}, {4, 5}, {5, 6}};
    Vec y = {2, 3, 4, 5, 6};

    LinearRegression lr;
    Vec theta = lr.fit(X, y);

    cout << "Final theta values: ";
    for (double t : theta) cout << t << " ";
    cout << endl;

    return 0;
}
